fairworkflows/fairstep.py

```diff
@@ -327,3 +356,78 @@ def __str__(self):
+    arguments to the decorator.
+
+    The raw code of the function will be used to set the description of the fair step.
+
+    The type annotations of the input arguments and return statement will be used to
+    automatically set inputs and outputs of the FAIR step.
+
+    Args:
+        label (str): Label of the fair step (corresponds to rdfs:label predicate)
+        is_pplan_step (str): Denotes whether this step is a pplan:Step
+        is_manual_task (str): Denotes whether this step is a bpmn.ManualTask
+        is_script_task (str): Denotes whether this step is a bpmn.ScriptTask
+
+    """
+    def _modify_function(func):
+        """
+        Store FairStep object as _fairstep attribute of the function. Use inspection to get the
+        description, inputs, and outputs of the step based on the function specification.
+        """
+        def _wrapper(*args, **kwargs):
+            return func(*args, **kwargs)
+        # Description of step is the raw function code
+        description = inspect.getsource(func)
+        inputs = _extract_inputs_from_function(func)
+        outputs = _extract_outputs_from_function(func)
+        _wrapper._fairstep = FairStep(label=label,
+                                      description=description,
+                                      is_pplan_step=is_pplan_step,
+                                      is_manual_task=is_manual_task,
+                                      is_script_task=is_script_task,
+                                      inputs=inputs,
+                                      outputs=outputs)
+        _wrapper._fairstep.validate()
+        return _wrapper
+    return _modify_function
+
+
+def _extract_inputs_from_function(func) -> List[FairVariable]:
+    """
+    Extract inputs from function using inspection. The name of the argument will be the name of
+    the fair variable, the corresponding type hint will be the type of the variable.
+    """
+    argspec = inspect.getfullargspec(func)
+    try:
+        return [FairVariable(name=arg, type=argspec.annotations[arg].__name__)
+                for arg in argspec.args]
+    except KeyError:
+        raise ValueError('Not all input arguments have type hinting,'
+                         'see https://docs.python.org/3/library/typing.html')
+
+
+def _extract_outputs_from_function(func) -> List[FairVariable]:
+    """
+    Extract outputs from function using inspection. The name will be {function_name}_output{
+    output_number}. The corresponding return type hint will be the type of the variable.
+    """
+    annotations = get_type_hints(func)
+    try:
+        return_annotation = annotations['return']
+    except KeyError:
+        raise ValueError('The return of the function does not have type hinting,'
+                         'see https://docs.python.org/3/library/typing.html')
+    # TODO: How to properly check that the return is annotated as Tuple?
```

Comment:
How to assert that the return is annotated as Tuple?
```python
type(return_annotation)
Out[2]: typing._GenericAlias
return_annotation
Out[3]: typing.Tuple[int, int]
return_annotation.__args__
Out[4]: (int, int)
```

Reply:
Lourens has a [solution for yatiml](https://github.com/yatiml/yatiml/blob/develop/yatiml/util.py#L28). Maybe I will leave it like this for now, I think this is buggy anyway... Or shall we add the `yatiml` dependency? (we need it for `nanopub` anyway).

Reply:
I'm totally fine with adding yatiml as a dependency.

Reply:
The `yatiml` solution didn't work, but I wrote a small piece of code that was influenced by it. It is more fool-proof now.